pkg/ingester/limiter_test.go

```diff
@@ -71,88 +71,125 @@ func runLimiterMaxFunctionTest(
-			shardSize:                0,
-			expectedValue:            300,
+		"zone-awareness disabled, limit is enabled with replication-factor=3 and the shard size > number of ingesters": {
+			globalLimit:           1000,
+			ringReplicationFactor: 3,
+			ringIngesterCount:     10,
+			shardSize:             20,  // Greater than number of ingesters.
+			expectedValue:         300, // (1000 / 10 healthy ingesters) * 3 replication factor
 		},
-		"zone-awareness enabled, limit enabled and the shard size is NOT divisible by number of zones": {
-			globalLimit:              900,
-			ringReplicationFactor:    3,
-			ringZoneAwarenessEnabled: true,
-			ringIngesterCount:        9,
-			ringZonesCount:           3,
-			shardSize:                5,   // Not divisible by number of zones.
-			expectedValue:            450, // (900 / 6) * 3
+
+		"zone-awareness enabled, limit is disabled": {
+			globalLimit:                 0,
+			ringReplicationFactor:       1,
+			ringIngesterCount:           1,
+			ringZoneAwarenessEnabled:    true,
+			ringZonesCount:              3,
+			healthyIngestersInZoneCount: 1,
+			expectedValue:               math.MaxInt32,
 		},
-		"zone-awareness enabled, limit enabled and the shard size is divisible by number of zones": {
-			globalLimit:              900,
-			ringReplicationFactor:    3,
-			ringZoneAwarenessEnabled: true,
-			ringIngesterCount:        9,
-			ringZonesCount:           3,
-			shardSize:                6,   // Divisible by number of zones.
-			expectedValue:            450, // (900 / 6) * 3
+		"zone-awareness enabled, limit is enabled with replication-factor=3, all ingesters up and running, and shard size 0": {
+			globalLimit:                 900,
+			ringReplicationFactor:       3,
+			ringIngesterCount:           9,
+			ringZoneAwarenessEnabled:    true,
+			ringZonesCount:              3,
+			healthyIngestersInZoneCount: 3, // 9 ingesters / 3 zones
+			shardSize:                   0,
+			expectedValue:               300, // (900 / 3 zones / 3 healthy ingesters per zone) * 3 replication factor
 		},
-		"zone-awareness enabled, limit enabled and the shard size > number of ingesters": {
-			globalLimit:              900,
-			ringReplicationFactor:    3,
-			ringZoneAwarenessEnabled: true,
-			ringIngesterCount:        9,
-			ringZonesCount:           3,
-			shardSize:                20, // Greater than number of ingesters.
-			expectedValue:            300,
+		"zone-awareness enabled, limit is enabled with replication-factor=3, some ingesters scaled down, and shard size 0": {
+			globalLimit:                 900,
+			ringReplicationFactor:       3,
+			ringIngesterCount:           9,
+			ringZoneAwarenessEnabled:    true,
+			ringZonesCount:              3,
+			healthyIngestersInZoneCount: 2, // in this zone ingesters are scaled down from 3 to 2
+			shardSize:                   0,
+			expectedValue:               450, // (900 / 3 zones / 2 healthy ingesters per zone) * 3 replication factor
```

Comment:
I am not sure I can answer this question. We would need @pracucci 's support. When I added the [HealthyInstancesInZoneCount()](https://github.com/grafana/dskit/blob/main/ring/lifecycler.go#L389) function to Lifecycler, I initially counted all instances, but then @pracucci suggested that we should count only the healthy ones, so I implemented it [this way](https://github.com/grafana/dskit/blob/main/ring/lifecycler.go#L819-L822).

Reply:
Regarding limits and healthy ingesters… we have a discussion whether using healthy ingesters only makes sense. Let’s consider an example. User has 1M series limit. We have 3 zones, each with 10 ingesters, and we use RF=3. So each ingester has 100k limit.

During rollout, we stop 5 ingesters at once in single zone. So there will be only 5 healthy ingesters left in the zone.
At this point limits at each remaining ingester in the same zone is recomputed to allow 200k series.
Mimir now always uses “WriteNoExtend”, so if user wasn’t hitting the 100k limit yet, this is fine. But if user was hitting 100k limit before, now we temporarily allow twice as many series for the user. After rollout in the zone finishes, per-ingester limit will be set back to 100k, but series will be in memory already.

Situation is similar when shuffle sharding is in use for the user. Let’s say user has 24 ingesters in total (ie. 8 per zone).
Limit per ingester if all ingesters are healhty will be 125k. During rollout we restart 5 ingesters in one zone, and only 5 will be healthy. Since 5 is less than 8 requested ingesters from the zone, temporary limit on each ingester will be 200k again.

Note that shuffle shard doesn’t take healthiness into consideration, so we’re mixing unhealthy with healthy ingesters when doing the check.
Given that we always use WriteNoExtend, I am not sure why we only consider healthy ingesters when computing the limit.

Reply:
Talked to Peter on Slack. I think the reason why we're not seeing it happening is because the ring client skips the entire zone if there are unhealthy instances, so we don’t write at all to ingesters in the zone we’re rollout out.

However, I'm 👍 with the change proposed by Peter, to count all instances not just healthy ones, given it looks more correct to me too.

Reply:
Does it mean we need to do the following things?

1. update `dskit` by replacing its current `HealthyInstancesInZoneCount()` by `InstancesInZoneCount()` and by updating its implementation
2. once it is merged, update this PR and use the new `InstancesInZoneCount()` method instead? Beside some syntactical changes (replacing "healthyIngesters" with "ingesters"), I don't see other things that should be changed here.

Reply:
That looks correct to me.

Reply:
I think you also need to add `InstancesCount()` in dskit to use when the zone-aware replication is disabled.

Reply:
@pstibrany @pracucci Could you please review https://github.com/grafana/dskit/pull/270?